### PR TITLE
Replace System trait with functions

### PR DESCRIPTION
**Pull request by msiglreith**
cc #7 

Splitting the current `exec` into two functions: `fetch` which will acquire the data bundle and `exec` which executes the worker function.
The main drawback is the introduction of more `Box` and the requirement of a `unsafe` block, wasn't able to remove this with `Any` due to `'static` restriction. (Helper helps!)

An new constraint is that `Resources` needs to outlive the `Dispatcher` but I think thats pretty minor..

### CODE REVIEW

File: src/dispatch.rs [hunk 1]
```diff
@@ -300,34 +311,11 @@ impl<'t> DispatcherBuilder<'t> {
...
-
-struct SystemInfo<'t> {
+struct SystemInfo {
     dependents: Vec<usize>,
-    exec: Box<ExecSystem + Send + 't>,
 }
+
+struct SystemWorker<'a> {
+    exec: Box<FnMut(Box<Helper + 'a>, &AtomicBitSet) + Send + Sync>,
+    fetch: Box<Fn(&'a Resources) -> Box<Helper + 'a> + Send + Sync>,
```
Comment (msiglreith) [inline src/dispatch.rs]: Adding the `Scope<'a>` into the closure makes the lifetime story quite difficult (impossible?).
That's why the fetch and exec part got split. The extra box is there to hide the return type.

**Comment by torkleyy** [comment]
@msiglreith Would it be possible to bind `FnMut`'s `SystemData` to a fake lifetime somehow?

**Comment by msiglreith** [comment]
@torkleyy Not really sure what you mean.
The current issue I'm facing is to satisfy the Scope lifetime somehow I think. Executing with spawning new threads works fine when combining the `exec` and `fetch` parts. Then it also looks similar to the current `System` trait based approach.

**Comment by torkleyy** [comment]
Well, `FnMut`s could work a lot like the `System` trait does now, if there was only a way to specify HRTBs for type parameters (which obviously does not make much sense). So if we could just use some fake lifetime, it would be like `where F: FnMut(D), D: SystemData<'fake>` and everything could be implemented just the way systems are. I don't feel very comfortable with this though.

> The current issue I'm facing is to satisfy the Scope lifetime

This was the part of `shred` I spent most of the time implementing it for.

**Comment by msiglreith** [comment]
Rethinking it a bit, and after implementing it for my old `respecs` proof of concept, the approach you mentioned above
> I think we might keep the System trait and have closures additionally. That way, you can use a closure if you don't need per-system data and you can use a struct if you do.

seems to be the way to go here. Fully closure based would be nice to have but seems not be possible without performance loss atm.
Thanks for the dicussion and taking the time!

**Pull request closed without merging**